File: solutions/python/nth-prime/1/nth_prime.py
from math import isqrt


def prime(n: int) -> int:
    if n < 1:
        raise ValueError("there is no zeroth prime")
    prime_iter = PrimeIter()
    for _ in range(n - 1):
        next(prime_iter)
    return next(prime_iter)


class PrimeIter:
    WHEEL = [4, 2, 4, 2, 4, 6, 2, 6]

    def __init__(self):
        self.primes = []
        self.next_candidate = 2
        self.wheel_idx = 0

    def is_prime(self, num: int) -> bool:
        if num % 2 == 0:
            return num == 2
        if num % 3 == 0:
            return num == 3
        if num % 5 == 0:
            return num == 5

        sqrt_num = isqrt(num)
        for p in self.primes:
            if p > sqrt_num:
                break
            if num % p == 0:
                return False
        return True

    def __iter__(self):
        return self

    def __next__(self) -> int:
        if len(self.primes) == 0:
            self.primes.append(2)
            self.next_candidate = 3
            return 2
        elif len(self.primes) == 1:
            self.primes.append(3)
            self.next_candidate = 5
            return 3
        elif len(self.primes) == 2:
            self.primes.append(5)
            self.next_candidate = 7
            return 5

        while not self.is_prime(self.next_candidate):
            self.next_candidate += self.WHEEL[self.wheel_idx]
            self.wheel_idx = (self.wheel_idx + 1) % 8

        prime = self.next_candidate
        self.primes.append(prime)
        self.next_candidate += self.WHEEL[self.wheel_idx]
        self.wheel_idx = (self.wheel_idx + 1) % 8
        return prime

File: solutions/typescript/nth-prime/1/nth-prime.ts
class PrimeIterator implements Iterator<number> {
    private static readonly WHEEL = [4, 2, 4, 2, 4, 6, 2, 6];
    #primes: number[] = [];
    #candidate = 2;
    #wheelPosition = 0;
    #cache = new Map<number, boolean>();

    #isPrime(n: number): boolean {
        if (this.#cache.has(n)) return this.#cache.get(n)!;

        const sqrt = Math.floor(Math.sqrt(n));
        for (let divisor = 2; divisor <= sqrt; divisor++) {
            if (n % divisor === 0) {
                this.#cache.set(n, false);
                return false;
            }
        }
        this.#cache.set(n, true);
        return true;
    }

    public next(): IteratorResult<number, number> {
        let value: number;

        if (this.#primes.length === 0) {
            this.#primes.push(2);
            this.#candidate = 3;
            value = 2;
        } else if (this.#primes.length === 1) {
            this.#primes.push(3);
            this.#candidate = 5;
            value = 3;
        } else if (this.#primes.length === 2) {
            this.#primes.push(5);
            this.#candidate = 7;
            value = 5;
        } else {
            while (!this.#isPrime(this.#candidate)) {
                this.#candidate += PrimeIterator.WHEEL[this.#wheelPosition];
                this.#wheelPosition = (this.#wheelPosition + 1) % 8;
            }

            value = this.#candidate;
            this.#primes.push(value);

            this.#candidate += PrimeIterator.WHEEL[this.#wheelPosition];
            this.#wheelPosition = (this.#wheelPosition + 1) % 8;
        }

        return { value, done: false };
    }
}

export function nth(n: number): number {
    if (n < 1) throw new Error("Prime is not possible");
    const primes = new PrimeIterator();

    for (let i = 0; i < n - 1; i++) primes.next();
    return primes.next().value;
}
